src/main/kotlin/com/cognifide/gradle/aem/common/instance/InstanceIMSClient.kt

```diff
@@ -0,0 +1,163 @@
+    }
+
+    private lateinit var privateKey: String
+
+    private lateinit var orgId: String
+
+    private lateinit var technicalAccountId: String
+
+    private lateinit var clientId: String
+
+    private lateinit var clientSecret: String
+
+    private lateinit var imsHost: String
+
+    private lateinit var imsExchangeEndpoint: String
+
+    private val expirationTime = System.currentTimeMillis() / 1000 + 86400L
+
+    private lateinit var metaScopes: List<String>
+
+    private fun readProperties() {
+        if (serviceTokenFile == null) {
+            throw InstanceException("No secret file available for generating access token")
+        }
+        val jsonString = Files.readString(serviceTokenFile!!.toPath())
+        val obj = JSONObject(jsonString)
+        val integration = obj.getJSONObject("integration")
+
+        privateKey = integration.getString("privateKey")
+        orgId = integration.getString("org")
+        technicalAccountId = integration.getString("id")
+        clientId = integration.getJSONObject("technicalAccount").getString("clientId")
+        clientSecret = integration.getJSONObject("technicalAccount").getString("clientSecret")
+        imsHost = integration.getString("imsEndpoint")
+        imsExchangeEndpoint = "https://$imsHost/ims/exchange/jwt"
+        metaScopes = integration.getString("metascopes").split(",")
+    }
+
+    private fun generateJWTToken(): String {
+        var privateKeyContent: ByteArray
+        PEMParser(StringReader(privateKey)).use { pemParser ->
+            val keyPair: KeyPair = JcaPEMKeyConverter().getKeyPair(pemParser.readObject() as PEMKeyPair)
+            privateKeyContent = keyPair.private.encoded
+        }
+        val keyFactory = KeyFactory.getInstance("RSA")
+        val keySpec: KeySpec = PKCS8EncodedKeySpec(privateKeyContent)
+        val rsaPrivateKey = keyFactory.generatePrivate(keySpec) as RSAPrivateKey
+
+        val jwtClaims = mutableMapOf<String, Any>(
+            "iss" to orgId,
+            "sub" to technicalAccountId,
+            "exp" to expirationTime,
+            "aud" to "https://$imsHost/c/$clientId",
+        )
+        val scopes = metaScopes.associate {
+            "https://$imsHost/s/$it" to true
+        }
+        jwtClaims.putAll(scopes)
+
+        return Jwts.builder()
+            .setClaims(jwtClaims)
+            .signWith(SignatureAlgorithm.RS256, rsaPrivateKey)
+            .compact()
+    }
+
+    private fun fetchAccessToken(jwtToken: String): String? {
+        val connection = URL(imsExchangeEndpoint).openConnection() as HttpsURLConnection
+        connection.requestMethod = HttpMethods.POST
+        val urlParameters = "client_id=$clientId&client_secret=$clientSecret&jwt_token=$jwtToken"
+
+        // Send post request
+        connection.doOutput = true
+        val writer = DataOutputStream(connection.outputStream)
+        writer.writeBytes(urlParameters)
+        writer.flush()
+        writer.close()
```

Comment:
for closeable in Kotlin you should use idiom `.use {}` 

https://kotlinlang.org/docs/idioms.html#java-7-s-try-with-resources